identity/aziot-identityd/src/lib.rs

```diff
@@ -702,76 +634,199 @@ impl Api {
+                            Some(attributes),
+                        )
+                        .await?;
+
+                    // Parse certificate expiration time.
+                    let expiration = get_cert_expiration(&certificate)?;
+
+                    aziot_identity_common::Identity::Local(aziot_identity_common::LocalIdSpec {
+                        private_key,
+                        certificate,
+                        expiration,
+                    })
+                }
+            };
+
+        Ok(local_identity)
+    }
+
     async fn create_identity_cert_if_not_exist_or_expired(
         &self,
-        identity_pk: &str,
+        identity_pk: KeyType,
         identity_cert: &str,
         subject: &str,
         attributes: Option<aziot_identity_common::LocalIdAttr>,
-    ) -> Result<(), Error> {
-        let device_id_cert = self.cert_client.get_cert(identity_cert).await;
-        let create_cert = match device_id_cert {
-            Ok(device_id_cert) => {
-                let x509_req = openssl::x509::X509::from_pem(&device_id_cert).map_err(|err| {
-                    Error::Internal(InternalError::CreateCertificate(Box::new(err)))
-                })?;
-                let cert_expiration = x509_req.as_ref().not_after();
-                let current_time = openssl::asn1::Asn1Time::days_from_now(0).map_err(|err| {
-                    Error::Internal(InternalError::CreateCertificate(Box::new(err)))
-                })?;
-                let expiration_time = current_time.diff(cert_expiration).map_err(|err| {
-                    Error::Internal(InternalError::CreateCertificate(Box::new(err)))
-                })?;
-                expiration_time.days < 1
-            }
-            Err(_) => true,
+    ) -> Result<(String, String), Error> {
+        // Retrieve existing cert and check it for expiry. Raw key bytes are not persisted,
+        // so a new cert must always be generated if KeyType is Raw.
+        let mut device_id_cert = match identity_pk {
+            KeyType::Raw => None,
+            KeyType::Handle(_) => match self.cert_client.get_cert(identity_cert).await {
+                Ok(pem) => {
+                    let cert = openssl::x509::X509::from_pem(&pem).map_err(|err| {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+                    let cert_expiration = cert.as_ref().not_after();
+                    let current_time =
+                        openssl::asn1::Asn1Time::days_from_now(0).map_err(|err| {
+                            Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                        })?;
+                    let expiration_time = current_time.diff(cert_expiration).map_err(|err| {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+
+                    if expiration_time.days < 1 {
+                        None
```

Comment:
We should log a warning that the identity certificates is being renewed to ensure > 24 hour expiry.

Reply:
Added. Note that I made it info level since I don't really think cert expiry (which is expected to happen everntually) should be a warning.